timemachine/fe/single_topology_v3.py

```diff
@@ -612,3 +612,95 @@ def U_fn(x):
             return (1 - lamb) * U0_fn(x) + lamb * U1_fn(x)
 
         return U_fn
+
+    def _setup_intermediate_bonded_term(self, src_bond, dst_bond, lamb, align_fn, interpolate_fn):
+
+        src_cls_bond = type(src_bond)
+        dst_cls_bond = type(dst_bond)
+
+        assert src_cls_bond == dst_cls_bond
+
+        bond_idxs_and_params = align_fn(
+            src_bond.get_idxs(),
+            src_bond.params,
+            dst_bond.get_idxs(),
+            dst_bond.params,
+        )
+        bond_idxs = []
+        bond_params = []
+        for idxs, src_params, dst_params in bond_idxs_and_params:
+            bond_idxs.append(idxs)
+            new_params = interpolate_fn(src_params, dst_params, lamb)
+            bond_params.append(new_params)
+
+        if src_cls_bond == potentials.NonbondedPairListPrecomputed:
+            assert src_bond.get_beta() == dst_bond.get_beta()
+            assert src_bond.get_cutoff() == dst_bond.get_cutoff()
+            return src_cls_bond(np.array(bond_idxs), src_bond.get_beta(), src_bond.get_cutoff()).bind(bond_params)
+        else:
+            return src_cls_bond(np.array(bond_idxs)).bind(bond_params)
+
+    def _setup_intermediate_chiral_bond_term(self, src_bond, dst_bond, lamb, interpolate_fn):
+
+        assert type(src_bond) == potentials.ChiralBondRestraint
+        assert type(dst_bond) == potentials.ChiralBondRestraint
+
+        idxs_and_params = interpolate.align_chiral_bond_idxs_and_params(
+            src_bond.get_idxs(),
+            src_bond.params,
+            src_bond.get_signs(),
+            dst_bond.get_idxs(),
+            dst_bond.params,
+            dst_bond.get_idxs(),
+        )
+        chiral_bond_idxs = []
+        chiral_bond_params = []
+        chiral_bond_signs = []
+        for idxs, sign, src_k, dst_k in idxs_and_params:
+            chiral_bond_idxs.append(idxs)
+            new_params = interpolate_fn(src_k, dst_k, lamb)
+            chiral_bond_params.append(new_params)
+            chiral_bond_signs.append(sign)
+
+        # these should be properly sized
+        chiral_bond_idxs = np.array(chiral_bond_idxs, dtype=np.int32).reshape((-1, 4))
+
+        return potentials.ChiralBondRestraint(chiral_bond_idxs, chiral_bond_signs).bind(chiral_bond_params)
+
+    def setup_intermediate_state(self, lamb):
+        """
+        Setup intermediate states at some value of lambda.
+        """
+        src_system = self.setup_end_state_src()
+        dst_system = self.setup_end_state_dst()
+        interpolate_fn = interpolate.linear_interpolation
```

Comment:
@mcwitt this is probably where you will supply more efficient interpolation schemes later on.